lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java

```diff
@@ -397,63 +460,68 @@ void onContainerRequest(final AMRMClient.ContainerRequest... containerRequests)
-        // It is okay to send nonzero m-capacity and n-capacity request together since bigger containers
-        // can be matched.
-        // TODO[JIRA REEF-42, REEF-942]: revisit this when implementing locality-strictness
-        // (i.e. a specific rack request can be ignored)
-        if (this.requestsAfterSentToRM.size() > 1) {
-          try {
-            this.resourceManager.removeContainerRequest(matchedRequest);
-          } catch (final Exception e) {
-            LOG.log(Level.WARNING, "Nothing to remove from Async AMRM client's queue, " +
-                "removal attempt failed with exception", e);
-          }
-        }
 
-        this.requestsAfterSentToRM.remove();
-        doHomogeneousRequests();
-
-        LOG.log(Level.FINEST, "Allocated Container: memory = {0}, core number = {1}",
-            new Object[]{container.getResource().getMemory(), container.getResource().getVirtualCores()});
-        this.reefEventHandlers.onResourceAllocation(ResourceEventImpl.newAllocationBuilder()
-            .setIdentifier(container.getId().toString())
-            .setNodeId(container.getNodeId().toString())
-            .setResourceMemory(container.getResource().getMemory())
-            .setVirtualCores(container.getResource().getVirtualCores())
-            .setRackName(rackNameFormatter.getRackName(container))
-            .setRuntimeName(RuntimeIdentifier.RUNTIME_NAME)
-            .build());
-        this.updateRuntimeStatus();
-      } else {
+      if (!matchContainerWithPendingRequest(container)) {
         LOG.log(Level.WARNING, "Got an extra container {0} that doesn't match, releasing...", container.getId());
         this.resourceManager.releaseAssignedContainer(container.getId());
+        return;
       }
+
+      final AMRMClient.ContainerRequest matchedRequest = this.requestsAfterSentToRM.peek();
+
+      this.containerRequestCounter.decrement();
+      this.containers.add(container);
+
+      LOG.log(Level.FINEST, "{0} matched with {1}", new Object[] {container, matchedRequest});
+
+      // Due to the bug YARN-314 and the workings of AMRMCClientAsync, when x-priority m-capacity zero-container
+      // request and x-priority n-capacity nonzero-container request are sent together, where m > n, RM ignores
+      // the latter.
+      // Therefore it is necessary avoid sending zero-container request, even it means getting extra containers.
+      // It is okay to send nonzero m-capacity and n-capacity request together since bigger containers
+      // can be matched.
+      // TODO[JIRA REEF-42, REEF-942]: revisit this when implementing locality-strictness
+      // (i.e. a specific rack request can be ignored)
+      if (this.requestsAfterSentToRM.size() > 1) {
+        try {
+          this.resourceManager.removeContainerRequest(matchedRequest);
+        } catch (final Exception e) {
+          LOG.log(Level.WARNING, "Nothing to remove from Async AMRM client's queue, " +
```

Comment:
Technically, anything can cause the exception, not only the empty async AMRM queue, so the first part of this log message seems a bit confusing.